WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift

```diff
@@ -51,23 +56,315 @@ import Foundation
+    /// - Returns: The updated post UploadOperation or nil
+    ///
+    fileprivate func combinePostAndMediaContentForGroupID(_ groupID: String) -> UploadOperation? {
+        guard let postUploadOp = fetchPostUploadOpForGroup(groupID),
+            let mediaUploadOps = fetchMediaUploadOpsForGroup(groupID) else {
+                return nil
+        }
+        let remoteURLText = mediaUploadOps.flatMap({ $0.remoteURL }).map({ "".stringByPrependingMediaURL($0) }).joined()
+        let content = postUploadOp.postContent ?? ""
+        postUploadOp.postContent = content + remoteURLText
+        coreDataStack.saveContext()
+
+        return postUploadOp
+    }
+
+    /// Uploads a post to the server
+    ///
+    /// - Parameter postUploadOp: The UploadOperation that represents a post
+    ///
+    fileprivate func uploadPost(postUploadOp: UploadOperation) {
+        let postUploadOpID = postUploadOp.objectID
+        guard let oauth2Token = ShareExtensionService.retrieveShareExtensionToken() else {
+            logError(errorString: "Error creating post: OAuth token is not defined.", uploadOpObjectIDs: postUploadOpID)
+            return
+        }
+        guard postUploadOp.siteID > 0 else {
+            logError(errorString: "Error creating post: site ID was missing.", uploadOpObjectIDs: postUploadOpID)
+            return
+        }
+
+        let api = WordPressComRestApi(oAuthToken: oauth2Token,
+                                      userAgent: nil,
+                                      backgroundUploads: false,
+                                      backgroundSessionIdentifier: backgroundSessionIdentifier,
+                                      sharedContainerIdentifier: WPAppGroupName)
+        let remote = PostServiceRemoteREST.init(wordPressComRestApi: api, siteID: NSNumber(value: postUploadOp.siteID))
+        postUploadOp.currentStatus = .InProgress
+        coreDataStack.saveContext()
+
+        remote.createPost(postUploadOp.remotePost, success: { post in
+            if let post = post {
+                if let postID = post.postID {
+                    DDLogInfo("Post \(post.postID.stringValue) sucessfully uploaded to site \(post.siteID.stringValue)")
+                    postUploadOp.remotePostID = postID.int64Value
+                }
+                postUploadOp.currentStatus = .Complete
+                self.coreDataStack.saveContext()
+                self.cleanupSessionAndTerminate()
+            }
+        }, failure: { error in
+            var errorString = "Error creating post"
+            if let error = error as NSError? {
+                errorString += ": \(error.localizedDescription)"
+            }
+            self.logError(errorString: errorString, uploadOpObjectIDs: postUploadOpID)
+            self.cleanupSessionAndTerminate()
+        })
+    }
 }
 
-extension ShareExtensionSessionManager: URLSessionTaskDelegate {
```

Comment:
I feel that all these methods related to the core data operations should move to a separate class something like ShareExtensionStore.

Reply:
Seems like a good plan to me. 👍